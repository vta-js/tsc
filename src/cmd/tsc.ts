import path from "path";
import chalk from "chalk";
import spawn from "../utils/spawn";
import copyFiles from "../utils/copyFiles";

export interface TscOptions {
  sourceDir: string;
  outDir: string;
  project: string;
  extensions: string[];
  exclude: string[];
  withBabel: boolean;
  esnext: boolean;
  cwd: string;
}
export default async function tsc(options: TscOptions, log: (msg: string) => void): Promise<void> {
  const sourceDir = path.relative(options.cwd, options.sourceDir);
  const outDir = options.withBabel ? options.outDir : path.relative(options.cwd, options.outDir);
  log(`tsc: build files from ${chalk.cyan(sourceDir)} to ${chalk.cyan(outDir)}`);
  await spawn(
    "tsc",
    [
      "-p",
      options.project,
      "--outDir",
      options.outDir,
      "--rootDir",
      options.sourceDir,
      "--newLine",
      "LF",
    ].concat(
      options.withBabel && options.esnext ? ["--module", "esnext", "--target", "esnext"] : [],
    ),
    {
      stdio: ["ignore", "ignore", "inherit"],
    },
  );
  log(`tsc: copy files from ${chalk.cyan(sourceDir)} to ${chalk.cyan(outDir)}`);
  await copyFiles(
    options.sourceDir,
    options.outDir,
    "**/*.*",
    (options.exclude || []).concat(
      options.extensions.length === 1
        ? [`**/*.${options.extensions[0].replace(/^\.+/, "")}`]
        : [`**/*.{${options.extensions.map(ext => ext.replace(/^\.+/, "")).join(",")}}`],
    ),
    options.sourceDir,
  );
}
