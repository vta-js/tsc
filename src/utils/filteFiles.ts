import path from "path";
import glob from "glob";

export default function filteFiles(
  pattern = "**/*.*",
  cwd: string = process.cwd(),
): Promise<string[]> {
  return new Promise((resolve, reject) => {
    glob(
      pattern,
      {
        cwd,
        realpath: true,
      },
      (err, matches) => {
        if (err) {
          reject(err);
        } else {
          resolve(matches.map(file => path.relative(cwd, file)));
        }
      },
    );
  });
}
