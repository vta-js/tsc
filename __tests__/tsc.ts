import path from "path";
import fse from "fs-extra";
import tsc from "../src";
import filteFiles from "../src/utils/filteFiles";

jest.setTimeout(100000);

describe("utils", () => {
  it("filteFiles", () => {
    return filteFiles("**/*.!(ts|tsx)", [], path.resolve(__dirname, "./data/project/src")).then(
      (files) => {
        expect(files.length).toBe(3);
      },
    );
  });
});

function testDistFile(file: string, expectedExists = true): Promise<void> {
  return fse.pathExists(file).then((exists) => {
    expect(exists).toBe(expectedExists);
    if (expectedExists) {
      return fse.readFile(file).then((content) => {
        expect(content.toString("utf-8")).toMatchSnapshot();
      });
    }
    return undefined;
  });
}

/* eslint-disable jest/expect-expect */
describe("tsc", () => {
  it("node-12.6.0", () =>
    tsc({
      project: "tsconfig-build.json",
      outDir: "../../__snapshots__/dist/node-12.6.0",
      extTs: ".ts,.tsx",
      extJs: ".js,.jsx",
      cwd: path.resolve(__dirname, "./data/project"),
    })
      .then(() =>
        testDistFile(path.resolve(__dirname, "__snapshots__/dist/node-12.6.0/utils/copyFiles.js")),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0/utils/copyFiles.d.ts"),
        ),
      ));

  it("node-12.6.0-no-dts", () =>
    tsc({
      project: "tsconfig-build-nodts.json",
      outDir: "../../__snapshots__/dist/node-12.6.0-no-dts",
      extTs: ".ts,.tsx",
      extJs: ".js,.jsx",
      cwd: path.resolve(__dirname, "./data/project"),
    })
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-dts/utils/copyFiles.js"),
        ),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-dts/utils/copyFiles.d.ts"),
          false,
        ),
      ));

  it("node-12.6.0-no-babel", () =>
    tsc({
      project: "tsconfig-build.json",
      babel: false,
      outDir: "../../__snapshots__/dist/node-12.6.0-no-babel",
      extTs: ".ts,.tsx",
      extJs: ".js,.jsx",
      cwd: path.resolve(__dirname, "./data/project"),
    })
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-babel/utils/copyFiles.js"),
        ),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-babel/utils/copyFiles.d.ts"),
        ),
      ));

  it("node-12.6.0-no-jsx", () =>
    tsc({
      project: "tsconfig-build-nodts.json",
      outDir: "../../__snapshots__/dist/node-12.6.0-no-jsx",
      extTs: ".ts,.tsx",
      extJs: ".js",
      cwd: path.resolve(__dirname, "./data/project"),
    })
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-jsx/utils/button.jsx"),
        ),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-jsx/utils/button.js"),
          false,
        ),
      ));

  it("node-12.6.0-exclude-jsx", () =>
    tsc({
      project: "tsconfig-build-nodts.json",
      outDir: "../../__snapshots__/dist/node-12.6.0-exclude-jsx",
      extTs: ".ts,.tsx",
      extJs: ".js",
      exclude: ["**/*.jsx"],
      cwd: path.resolve(__dirname, "./data/project"),
    })
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-exclude-jsx/utils/button.jsx"),
          false,
        ),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-exclude-jsx/utils/button.js"),
          false,
        ),
      ));

  it("node-12.6.0-no-tsx", () =>
    tsc({
      project: "tsconfig-build-notsx.json",
      outDir: "../../__snapshots__/dist/node-12.6.0-no-tsx",
      extTs: ".ts",
      extJs: ".js,.jsx",
      cwd: path.resolve(__dirname, "./data/project"),
    })
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-tsx/utils/helper/button.tsx"),
        ),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-tsx/utils/filteFiles.ts"),
          false,
        ),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-12.6.0-no-tsx/utils/helper/button.js"),
          false,
        ),
      ));

  it("node-8.16.0", () =>
    tsc({
      project: "tsconfig-build.json",
      babel: "babel.config-8.16.0.js",
      outDir: "../../__snapshots__/dist/node-8.16.0",
      extTs: ".ts,.tsx",
      extJs: ".js,.jsx",
      cwd: path.resolve(__dirname, "./data/project"),
    }).then(() =>
      testDistFile(path.resolve(__dirname, "__snapshots__/dist/node-8.16.0/utils/copyFiles.js")),
    ));

  it("node-8.16.0-no-esnext", () =>
    tsc({
      project: "tsconfig-build.json",
      babel: "babel.config-8.16.0.js",
      outDir: "../../__snapshots__/dist/node-8.16.0-no-esnext",
      extTs: ".ts,.tsx",
      extJs: ".js,.jsx",
      cwd: path.resolve(__dirname, "./data/project"),
      esnext: false,
    })
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-8.16.0-no-esnext/utils/copyFiles.js"),
        ),
      )
      .then(() =>
        testDistFile(
          path.resolve(__dirname, "__snapshots__/dist/node-8.16.0-no-esnext/utils/copyFiles.d.ts"),
        ),
      ));

  it("node-4.0.0", () =>
    tsc({
      project: "tsconfig-build.json",
      babel: "babel.config-4.0.0.js",
      outDir: "../../__snapshots__/dist/node-4.0.0",
      extTs: ".ts,.tsx",
      extJs: ".js,.jsx",
      cwd: path.resolve(__dirname, "./data/project"),
      silent: true,
    }).then(() =>
      testDistFile(path.resolve(__dirname, "__snapshots__/dist/node-4.0.0/utils/copyFiles.js")),
    ));

  it("fail-tsconfig", () =>
    tsc({
      project: "tsconfig-fail.json",
      outDir: "../../__snapshots__/dist/fail-tsconfig",
      cwd: path.resolve(__dirname, "./data/project-fail"),
    }).then((err) => {
      expect(!!err).toBe(true);
      expect(err.message.indexOf("Cannot use JSX unless the '--jsx' flag is provided") >= 0).toBe(
        true,
      );
    }));

  it("fail-babel", () =>
    tsc({
      babel: "babel-fail.config.js",
      outDir: "../../__snapshots__/dist/fail-babel",
      cwd: path.resolve(__dirname, "./data/project-fail"),
    }).then((err) => {
      expect(!!err).toBe(true);
      expect(err.message.indexOf("Cannot find module '@vta/tsc-invalid-babel-plugin'") >= 0).toBe(
        true,
      );
    }));
});
