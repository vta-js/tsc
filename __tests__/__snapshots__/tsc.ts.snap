// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tsc node-4.0.0 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

require(\\"core-js/modules/es.array.slice\\");

require(\\"core-js/modules/es.object.to-string\\");

require(\\"core-js/modules/es.promise\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = copyFiles;

var _regenerator = _interopRequireDefault(require(\\"@babel/runtime/regenerator\\"));

require(\\"regenerator-runtime/runtime\\");

var _asyncToGenerator2 = _interopRequireDefault(require(\\"@babel/runtime/helpers/asyncToGenerator\\"));

var _slicedToArray2 = _interopRequireDefault(require(\\"@babel/runtime/helpers/slicedToArray\\"));

var _path = _interopRequireDefault(require(\\"path\\"));

var _fsExtra = _interopRequireDefault(require(\\"fs-extra\\"));

var _filteFiles = _interopRequireDefault(require(\\"./filteFiles\\"));

function copyFilesOneByOne(from, dest, files) {
  if (files.length > 0) {
    var _files = (0, _slicedToArray2.default)(files, 1),
        file = _files[0];

    var destFile = _path.default.resolve(dest, file);

    return _fsExtra.default.ensureDir(_path.default.dirname(destFile)).then(function () {
      return _fsExtra.default.copyFile(_path.default.resolve(from, file), destFile);
    }).then(function () {
      return copyFilesOneByOne(from, dest, files.slice(1));
    });
  }

  return Promise.resolve();
}

function copyFiles(_x, _x2, _x3, _x4) {
  return _copyFiles.apply(this, arguments);
}

function _copyFiles() {
  _copyFiles = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(from, dest, pattern, cwd) {
    var files;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (0, _filteFiles.default)(pattern, cwd);

          case 2:
            files = _context.sent;
            _context.next = 5;
            return copyFilesOneByOne(from, dest, files);

          case 5:
          case \\"end\\":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _copyFiles.apply(this, arguments);
}"
`;

exports[`tsc node-8.16.0 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

require(\\"core-js/modules/es.array.iterator\\");

require(\\"core-js/modules/es.promise\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = copyFiles;

var _path = _interopRequireDefault(require(\\"path\\"));

var _fsExtra = _interopRequireDefault(require(\\"fs-extra\\"));

var _filteFiles = _interopRequireDefault(require(\\"./filteFiles\\"));

function copyFilesOneByOne(from, dest, files) {
  if (files.length > 0) {
    const [file] = files;

    const destFile = _path.default.resolve(dest, file);

    return _fsExtra.default.ensureDir(_path.default.dirname(destFile)).then(() => _fsExtra.default.copyFile(_path.default.resolve(from, file), destFile)).then(() => copyFilesOneByOne(from, dest, files.slice(1)));
  }

  return Promise.resolve();
}

async function copyFiles(from, dest, pattern, cwd) {
  const files = await (0, _filteFiles.default)(pattern, cwd);
  await copyFilesOneByOne(from, dest, files);
}"
`;

exports[`tsc node-8.16.0-no-esnext 1`] = `
"\\"use strict\\";

require(\\"core-js/modules/es.array.iterator\\");

require(\\"core-js/modules/es.promise\\");

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator[\\"throw\\"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

var __importDefault = void 0 && (void 0).__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    \\"default\\": mod
  };
};

define([\\"require\\", \\"exports\\", \\"path\\", \\"fs-extra\\", \\"./filteFiles\\"], function (require, exports, path_1, fs_extra_1, filteFiles_1) {
  \\"use strict\\";

  Object.defineProperty(exports, \\"__esModule\\", {
    value: true
  });
  path_1 = __importDefault(path_1);
  fs_extra_1 = __importDefault(fs_extra_1);
  filteFiles_1 = __importDefault(filteFiles_1);

  function copyFilesOneByOne(from, dest, files) {
    if (files.length > 0) {
      const [file] = files;
      const destFile = path_1.default.resolve(dest, file);
      return fs_extra_1.default.ensureDir(path_1.default.dirname(destFile)).then(() => fs_extra_1.default.copyFile(path_1.default.resolve(from, file), destFile)).then(() => copyFilesOneByOne(from, dest, files.slice(1)));
    }

    return Promise.resolve();
  }

  function copyFiles(from, dest, pattern, cwd) {
    return __awaiter(this, void 0, void 0, function* () {
      const files = yield filteFiles_1.default(pattern, cwd);
      yield copyFilesOneByOne(from, dest, files);
    });
  }

  exports.default = copyFiles;
});"
`;

exports[`tsc node-8.16.0-no-esnext 2`] = `
"export default function copyFiles(from: string, dest: string, pattern: string, cwd: string): Promise<void>;
"
`;

exports[`tsc node-12.6.0 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = copyFiles;

var _path = _interopRequireDefault(require(\\"path\\"));

var _fsExtra = _interopRequireDefault(require(\\"fs-extra\\"));

var _filteFiles = _interopRequireDefault(require(\\"./filteFiles\\"));

function copyFilesOneByOne(from, dest, files) {
  if (files.length > 0) {
    const [file] = files;

    const destFile = _path.default.resolve(dest, file);

    return _fsExtra.default.ensureDir(_path.default.dirname(destFile)).then(() => _fsExtra.default.copyFile(_path.default.resolve(from, file), destFile)).then(() => copyFilesOneByOne(from, dest, files.slice(1)));
  }

  return Promise.resolve();
}

async function copyFiles(from, dest, pattern, cwd) {
  const files = await (0, _filteFiles.default)(pattern, cwd);
  await copyFilesOneByOne(from, dest, files);
}"
`;

exports[`tsc node-12.6.0 2`] = `
"export default function copyFiles(from: string, dest: string, pattern: string, cwd: string): Promise<void>;
"
`;

exports[`tsc node-12.6.0-no-babel 1`] = `
"var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator[\\"throw\\"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { \\"default\\": mod };
};
define([\\"require\\", \\"exports\\", \\"path\\", \\"fs-extra\\", \\"./filteFiles\\"], function (require, exports, path_1, fs_extra_1, filteFiles_1) {
    \\"use strict\\";
    Object.defineProperty(exports, \\"__esModule\\", { value: true });
    path_1 = __importDefault(path_1);
    fs_extra_1 = __importDefault(fs_extra_1);
    filteFiles_1 = __importDefault(filteFiles_1);
    function copyFilesOneByOne(from, dest, files) {
        if (files.length > 0) {
            const [file] = files;
            const destFile = path_1.default.resolve(dest, file);
            return fs_extra_1.default
                .ensureDir(path_1.default.dirname(destFile))
                .then(() => fs_extra_1.default.copyFile(path_1.default.resolve(from, file), destFile))
                .then(() => copyFilesOneByOne(from, dest, files.slice(1)));
        }
        return Promise.resolve();
    }
    function copyFiles(from, dest, pattern, cwd) {
        return __awaiter(this, void 0, void 0, function* () {
            const files = yield filteFiles_1.default(pattern, cwd);
            yield copyFilesOneByOne(from, dest, files);
        });
    }
    exports.default = copyFiles;
});
"
`;

exports[`tsc node-12.6.0-no-babel 2`] = `
"export default function copyFiles(from: string, dest: string, pattern: string, cwd: string): Promise<void>;
"
`;

exports[`tsc node-12.6.0-no-dts 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = copyFiles;

var _path = _interopRequireDefault(require(\\"path\\"));

var _fsExtra = _interopRequireDefault(require(\\"fs-extra\\"));

var _filteFiles = _interopRequireDefault(require(\\"./filteFiles\\"));

function copyFilesOneByOne(from, dest, files) {
  if (files.length > 0) {
    const [file] = files;

    const destFile = _path.default.resolve(dest, file);

    return _fsExtra.default.ensureDir(_path.default.dirname(destFile)).then(() => _fsExtra.default.copyFile(_path.default.resolve(from, file), destFile)).then(() => copyFilesOneByOne(from, dest, files.slice(1)));
  }

  return Promise.resolve();
}

async function copyFiles(from, dest, pattern, cwd) {
  const files = await (0, _filteFiles.default)(pattern, cwd);
  await copyFilesOneByOne(from, dest, files);
}"
`;

exports[`tsc node-12.6.0-no-jsx 1`] = `
"import React from \\"react\\";

console.log(\\"utils.button\\");

export function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export default ({ text: propText }) => {
  const [text, setText] = React.useState(propText);
  React.useEffect(() => {
    setText(propText);
  }, [propText]);

  return <button>{text}</button>;
};
"
`;

exports[`tsc node-12.6.0-no-tsx 1`] = `
"import React from \\"react\\";

console.log(\\"utils.helper.button\\");

export function delay(ms): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export default ({ text: propText }: { text: string }): React.ReactNode => {
  const [text, setText] = React.useState(propText);
  React.useEffect(() => {
    setText(propText);
  }, [propText]);

  return <button>{text}</button>;
};
"
`;
